### PR TITLE
Nano : Enhancements for InferenceOptimizer to support YOLOX

### PR DESCRIPTION
**Pull request by rnwang04**
## Description
To support applying InferenceOptimizer on YOLOX.

### 1. Why the change?

Based on https://github.com/intel-analytics/BigDL/issues/5995, we want to extend `metric` in `InferenceOptimizer.optimize` to support more cases.

### 2. User API changes

Before API supports torchmetrics.Metric object or similar callable object which which takes prediction and target as input.
case1 : torchmetrics.Metric object
```python
inference_opt.optimize(model=model,
                       training_data=train_loader,
                       validation_data=val_loader,
                       metric=torchmetrics.Accuracy(),
                       direction="max",
                       thread_num=1)
```
case2：similar callable object which which takes prediction and target as input. related example: https://github.com/intel-analytics/BigDL/blob/main/python/nano/example/pytorch/inference_optimizer/resnet/inference_pipeline.py
```python
def metric(pred, target):
      value = process(pred, target)
      return value

inference_opt.optimize(model=model,
                       training_data=train_loader,
                       validation_data=val_loader,
                       metric=metric,
                       direction="max",
                       thread_num=1)
```
Now, two additional metric inputs are supported.
case3: A callable object that takes model and validation_data as input
```python
def metric(model, data_loader):
      value = process(model, data_loader)
      return value

inference_opt.optimize(model=model,
                       training_data=train_loader,
                       validation_data=val_loader,
                       metric=metric,
                       direction="max",
                       thread_num=1)
```
case4: A callable object that takes model as input, here validation_data is None, related example: https://github.com/intel-analytics/BigDL/issues/5995
```python
def metric(model):
      value = process(model)
      return value

inference_opt.optimize(model=model,
                       training_data=train_loader,
                       validation_data=None,
                       metric=metric,
                       direction="max",
                       thread_num=1)
```

### 3. Summary of the change

- [x] Extend metric to support more cases
- [x] Add more unittests to test different kinds of metric

### 4. How to test?
- [x] Unit test
- [x] Application test

Inference result of YOLOX：
image

### CODE REVIEW

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 1]
```diff
@@ -139,8 +139,15 @@ def optimize(self, model: nn.Module,
...
-               and returns a accuracy value in this calling method `metric(pred, target)`
+        :param metric: (optional) A callable object which is used for calculating accuracy.
+               1. If validation_data is composed of (input_data, target), metric could be
+               a torchmetrics.metric.Metric or similar object or function which takes
+               prediction and target then returns an accuracy value in this
+               calling method `accuracy = metric(pred, target)`.
+               2. Otherwise you can define a custom metric in this calling
+               method `accuracy = metric(model, batch_data)`.
+               3. Metric can even only recieve model as input in this calling
+               method `accuracy = metric(model)`, like coco_evaluator.
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:150]: The logic seems a little bit messy. Maybe it's simpler if we just require the metric be a callable object that takes model and data_loader (if not None) as input.

We can provide a helper class such as `NanoMetric` to help user compute the result if the user's metric takes (input_data, target), similar to what `_accuracy_calculate_helper` does; the user can then simply wrap their metric using `NanoMetric`

In addition, we can also automatically wrap the metric using `NanoMetric` inside `optimize` if it is a `TorchMetrics` object.

Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py]: Thanks for such great suggestions. I have modified code according to these comments.  Could you please take a look again to confirm that whether the current implementation is consistent with your vision ? @jason-dai 

Another question is, here I just ask user to provide a callable object that takes model and dataloader as input. Should I impose certain restrictions on this behavior, for example, given a Base class and ask user to inherit this class to implement custom logic?
```python
class BaseMetric(metaclass=ABCMeta):
    def __init__(self, metric: Callable):
        self.metric = metric
    
    @abstractmethod
    def __call__(self, model: nn.Module, data_loader: DataLoader = None):
        pass

class CustomMetric(BaseMetric):
    def __init__(self, metric: Callable):
        self.metric = metric
    
    def __call__(self, model: nn.Module, data_loader: DataLoader = None):
        # custom logic
        pass

metric = Accuracy()
metric = CustomMetric(metric)
```

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 2]
```diff
@@ -285,9 +295,22 @@ def func_test(model, input_sample):
...
+                        if method == "original":
+                            # test whether metric works
+                            try:
+                                result_map[method]["accuracy"] =\
+                                    _accuracy_calculate_helper(acce_model, metric,
+                                                               validation_data)
+                            except Exception:
+                                self._calculate_accuracy = False
+                                invalidInputError(False,
+                                                  "Your metric is incompatible with "
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:307]: explain what's the expected metric pattern in the error message

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 3]
```diff
@@ -750,17 +776,35 @@ def _throughput_calculate_helper(iterrun, baseline_time, func, *args):
...
+    import inspect
+    sig = inspect.signature(function)
+    if len(sig.parameters.values()) != 2:
+        return False
+    param1_name = list(sig.parameters.values())[0].name
+    param2_name = list(sig.parameters.values())[0].name
+    if "pred" in param1_name or "pred" in param2_name:
+        return True
+    if "target" in param2_name or "target" in param2_name:
+        return True
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:793]: does this make sense?

Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:793]: here we assume user's custom metric function is like this ` function_name(pred=xxx, target=xxx)`, a example is : https://github.com/intel-analytics/BigDL/blob/main/python/nano/example/pytorch/inference_optimizer/resnet/inference_pipeline.py

I think we have to limit the function signature here, otherwise we can't distinguish metric(pred, target) from metric (model, data_loader).

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 4]
```diff
@@ -139,8 +141,15 @@ def optimize(self, model: nn.Module,
...
-               and returns a accuracy value in this calling method `metric(pred, target)`
+        :param metric: (optional) A callable object which is used for calculating accuracy.
+               It supports two kinds of callable object:
+               1. A torchmetrics.Metric object or similar callable object which takes
+               prediction and target than returns an accuracy value in this calling
+               method `metric(pred, target)`. This requires data in validation_data
+               is composed of (input_data, target).
+               2. A callable object that takes model and validation_data(which could
+               be None) as input and returns an accuracy value in this calling method
+               `accuracy = metric(model, data_loader)`.
```
[the remark this answers is not in the document]
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:152]: done

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py (hunk 2 above)
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py]: done

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 5]
```diff
@@ -750,17 +784,35 @@ def _throughput_calculate_helper(iterrun, baseline_time, func, *args):
...
+    import inspect
+    sig = inspect.signature(function)
+    if len(sig.parameters.values()) != 2:
+        return False
+    param1_name = list(sig.parameters.values())[0].name
+    param2_name = list(sig.parameters.values())[0].name
+    if "pred" in param1_name:
+        return True
+    if "target" in param2_name:
+        return True
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:801]: `if "pred" in param1_name and "target" in param2_name`, yes?

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 6]
```diff
@@ -750,17 +784,35 @@ def _throughput_calculate_helper(iterrun, baseline_time, func, *args):
...
-    metric_list = []
-    sample_num = 0
-    with torch.no_grad():
-        for i, (data_input, target) in enumerate(data):
-            metric_list.append(metric(model(data_input), target).numpy() * data_input.shape[0])
-            sample_num += data_input.shape[0]
-    return np.sum(metric_list) / sample_num
+    if isinstance(metric, Metric) or _signature_check(metric) is True:
+        metric = NanoMetric(metric)
+        return metric(model, data)
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:811]: what if `data` is None?

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py [hunk 7]
```diff
@@ -750,17 +784,35 @@ def _throughput_calculate_helper(iterrun, baseline_time, func, *args):
...
     A quick helper to calculate accuracy
     '''
-    metric_list = []
-    sample_num = 0
-    with torch.no_grad():
-        for i, (data_input, target) in enumerate(data):
-            metric_list.append(metric(model(data_input), target).numpy() * data_input.shape[0])
-            sample_num += data_input.shape[0]
-    return np.sum(metric_list) / sample_num
+    if isinstance(metric, Metric) or _signature_check(metric) is True:
```
Comment (jason-dai) [inline python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:809]: What if data is None?

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py (hunk 5 above)
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:801]: will change to this.

File: python/nano/src/bigdl/nano/pytorch/inference/optimizer.py (hunk 7 above)
Comment (rnwang04) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:809]: will add ` invalidInputError(data is not None,xxx)` for this case.
Error message will look like this:
image
Do you think it's ok ?

Comment (jason-dai) [reply python/nano/src/bigdl/nano/pytorch/inference/optimizer.py:808]: ok
